modules/pricer/src/main/java/com/opengamma/strata/pricer/impl/volatility/smile/function/VolatilityFunctionProvider.java

```diff
@@ -56,6 +56,35 @@ public ValueDerivatives getVolatilityAdjoint(double forward, double strike, doub
     return ValueDerivatives.of(volatility, DoubleArray.ofUnsafe(res));
   }
 
+  /**
+   * Computes the first and second order derivatives of the volatility. 
+   * <p>
+   * The first derivative values will be stored in the input array {@code volatilityD} 
+   * The array contains, [0] Derivative w.r.t the forward, [1] the derivative w.r.t the strike, then followed by model 
+   * parameters. 
+   * Thus the length of the array should be 2 + (number of model parameters).  
+   * <p>
+   * The second derivative values will be stored in the input array {@code volatilityD2}. 
+   * Only the second order derivative with respect to the forward and strike must be implemented.
+   * The array contains [0][0] forward-forward; [0][1] forward-strike; [1][1] strike-strike.
+   * Thus the size should be 2 x 2.
+   * 
+   * @param forward  the forward value of the underlying
+   * @param strike  the strike value of the option
+   * @param timeToExpiry  the time to expiry of the option
+   * @param data The model data.
+   * @param volatilityD  the array used to return the first order derivative. 
+   * @param volatilityD2 The array of array used to return the second order derivative. 
+   * @return the volatility.
+   */
+  public abstract double getVolatilityAdjoint2(
```

Comment:
This is a very complex API. Could we use `ValueDerivatives` ? Could the 2D array be replaced with a 1D array as it only has three known values? In general, it is not good to mutate array parameters.